### PR TITLE
Clean the Trainer state

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PRs clean the fields used inside the `Trainer` to store state and gather them all in a clear class named `TrainerState` that is typed so the user knows exactly what they can access when subclassing and overriding methods (and for my next step, when writing callbacks).

As a result, the `log_history` does not need to be saved and loaded separately and the hack to get the steps we were at from the checkpoint folder name is removed (the user can copy all the checkpoint saved by the `Trainer` in a previous training in any folder and use it for resuming training). This PRs adds a test of the full reproducibility of a training resumed from a checkpoint.

There is a tiny breaking change that would affect users that trained a model using an earlier version of transformers and would like to resume it with a version obtained after this commit but I don't think this matters much.

It also enforces that the `TrainingArguments` passed to a `Trainer` are not changed during training, to avoid subtle bugs when launching several trainings in a row. This is verified by a new test.

### CODE REVIEW

**Comment by sgugger** [comment]
Tests pass in a multi-GPU environment and the specific `test_distributed_trainer` passes too. There is just one test that requires a batch size not too big so I manually skip it if there are more than 2 GPUs.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -260,10 +259,11 @@ def __init__(
...
-        self.log_history = []
         if "prediction_loss_only" in kwargs:
             warnings.warn(
-                "Passing `prediction_loss_only` as a keyword argument is deprecated and won't be possible in a future version. Use `args.prediction_loss_only` instead.",
+                "Passing `prediction_loss_only` as a keyword argument is deprecated and won't be possible in a "
+                + "future version. Use `args.prediction_loss_only` instead. Setting "
+                + f"`args.prediction_loss_only={kwargs['prediction_loss_only']}",
                 FutureWarning,
             )
             self.args.prediction_loss_only = kwargs.pop("prediction_loss_only")
```
Comment (sgugger) [inline src/transformers/trainer.py:269]: This is the only instance where the TrainingArguments are modified, with a clear warning.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1020,24 +1016,17 @@ def log(self, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
...
-        if self.epoch is not None:
-            logs["epoch"] = self.epoch
-        if self.total_flos is not None:
-            if self.args.local_rank != -1:
-                total_flos = distributed_broadcast_scalars([self.total_flos]).sum().item()
-            else:
-                total_flos = self.total_flos
-            if total_flos > 0:
-                logs["total_flos"] = total_flos
-        if self.global_step is None:
```
Comment (sgugger) [inline src/transformers/trainer.py:1034]: This is not necessary anymore since `global_step` is initialized to 0.

**Review by LysandreJik — APPROVED** [review]
I think this is a great addition to help clarify the inner workings of the `Trainer`.

This change brings a very different setup to what is implemented in the `TFTrainer`, and we would like to keep the two trainers as aligned as possible.

@jplu, what do you think of this change? It doesn't need to be implemented in the `TFTrainer` immediately, but having your approval for such a trainer-related change is important. Could the two trainers eventually leverage the same `TrainerState` for internal management?

**Comment by jplu** [comment]
Nice PR!!!

I think it is a nice addition. Only the `global_step` won't be necessary as it is already integrated by default into the TF checkpoints. This is perfect!

I plan also to add Keras callbacks into the TF Trainer that can handles few more arguments that are in this "state" class such as best checkpoint, and metrics.

It won't change much in the TF Trainer and could be very easily integrated as most of the "state" arguments work the same way for both Trainers. We can clearly imagine this state class for a much broader usage.

**Review by julien-c — APPROVED** [review]

**Comment by sgugger** [comment]
Checked the tests on TPU are passing, so can safely merge this.

**Pull request merged**